### PR TITLE
gateio: websocket updates

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

* Optimise unmarshal processes in some functions by using json.Rawmessage field. Prior was unmarshalling the same thing 3 times
* Fix processing of `spotOrderbookTickerChannel`
* Streamlined event response matching 
* Reduced allocations in `processOrderbookUpdate` and `processOrderbookSnapshot`
* Assign asset type to `stream.ChannelSubscription` type which was missing
* Batch outbound orderbook ticker so that every asset can be subscribed to immediately without initial sync interruption

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/gateio/gateio_websocket.go [hunk 1]
```diff
@@ -863,14 +852,30 @@ func (g *Gateio) generatePayload(event string, channelsToSubscribe []stream.Chan
...
 			Event:   event,
 			Channel: channelsToSubscribe[i].Channel,
 			Payload: params,
 			Auth:    auth,
 			Time:    timestamp.Unix(),
 		}
+
+		if channelsToSubscribe[i].Channel == "spot.book_ticker" {
+			// So to get all assets subscribed it needs to be batched and only
+			// spot.book_ticker can be batched, if not it will take about half
```
Comment (gloriousCode) [inline exchanges/gateio/gateio_websocket.go:867]: OBJECTION!!! `spot.tickers` can be batched too https://www.gate.io/docs/developers/apiv4/ws/en/#tickers-channel

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Big fan of not unmarshalling the same thing over and over, lots of good little fixes here

File: exchanges/gateio/gateio_websocket.go [hunk 2]
```diff
@@ -304,32 +294,46 @@ func (g *Gateio) processCandlestick(data []byte) error {
...
+func (g *Gateio) processOrderbookTicker(incoming []byte) error {
+	var data WsOrderbookTickerData
+	err := json.Unmarshal(incoming, &data)
 	if err != nil {
 		return err
 	}
-	g.Websocket.DataHandler <- tickerData
-	return nil
+
+	avail, err := g.GetAvailablePairs(asset.Spot)
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/gateio_websocket.go:304]: WOW! that's not ideal 😆Nice find! 🦅

File: exchanges/gateio/gateio_websocket.go (hunk 1 above)
Comment (shazbert) [reply exchanges/gateio/gateio_websocket.go:867]: I was just meaning the orderbooks, but I see your pointerino.

File: exchanges/gateio/gateio_ws_futures.go [hunk 3]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
-	response.Result = update
-	err := json.Unmarshal(data, &response)
+func (g *Gateio) processFuturesAndOptionsOrderbookUpdate(incoming []byte, assetType asset.Item) error {
+	var data WsFuturesAndOptionsOrderbookUpdate
+	err := json.Unmarshal(incoming, &data)
 	if err != nil {
 		return err
 	}
-	pair, err := currency.NewPairFromString(update.ContractName)
+	pair, err := currency.NewPairFromString(data.ContractName)
```
Comment (gloriousCode) [inline exchanges/gateio/gateio_ws_futures.go:530]: I'm not sure exactly whats causing it, but actions tests are showing 
` gateio_test.go:2900: GateIO websocket options orderbook update push data error: cannot create currency pair from  string too short to be a currency pair`
and `data` is an empty struct at this point

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Only one thing has come up for me with this round of testing

File: exchanges/gateio/gateio_ws_futures.go (hunk 3 above)
Comment (shazbert) [reply exchanges/gateio/gateio_ws_futures.go:530]: 🦅

**Review by gloriousCode — APPROVED** [review]
image
Thanks for the fixes and this PR!

**Review by thrasher- — CHANGES_REQUESTED** [review]
Nice improvements!

File: exchanges/gateio/gateio_ws_futures.go [hunk 4]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
 	updates := orderbook.Update{
-		UpdateTime: time.UnixMilli(update.TimestampInMs),
+		UpdateTime: time.UnixMilli(data.TimestampInMs),
 		Pair:       pair,
 		Asset:      assetType,
 	}
-	updates.Bids = make([]orderbook.Item, len(update.Bids))
-	updates.Asks = make([]orderbook.Item, len(update.Asks))
+	updates.Asks = make([]orderbook.Item, len(data.Asks))
 	for x := range updates.Asks {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/gateio_ws_futures.go:559]: 🦅

File: exchanges/gateio/gateio_ws_futures.go [hunk 5]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
-		updates.Asks[x] = orderbook.Item{
-			Amount: update.Asks[x].Size,
-			Price:  update.Asks[x].Price,
-		}
+		updates.Asks[x].Amount = data.Asks[x].Size
+		updates.Asks[x].Price = data.Asks[x].Price
 	}
+	updates.Bids = make([]orderbook.Item, len(data.Bids))
 	for x := range updates.Bids {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/gateio_ws_futures.go:564]: 🦅

File: exchanges/gateio/gateio_websocket.go [hunk 6]
```diff
@@ -347,43 +344,33 @@ func (g *Gateio) processOrderbookUpdate(data []byte) error {
...
-		UpdateTime: update.UpdateTimeMs.Time(),
+		UpdateTime: data.UpdateTimeMs.Time(),
 		Pair:       pair,
 	}
-	updates.Bids = make([]orderbook.Item, len(update.Bids))
-	updates.Asks = make([]orderbook.Item, len(update.Asks))
-	var price float64
-	var amount float64
+	updates.Asks = make([]orderbook.Item, len(data.Asks))
 	for x := range updates.Asks {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_websocket.go:354]:
```suggestion
	for x := range data.Asks {
```

File: exchanges/gateio/gateio_websocket.go [hunk 7]
```diff
@@ -347,43 +344,33 @@ func (g *Gateio) processOrderbookUpdate(data []byte) error {
...
 		if err != nil {
 			return err
 		}
-		updates.Asks[x] = orderbook.Item{
-			Amount: amount,
-			Price:  price,
-		}
 	}
+	updates.Bids = make([]orderbook.Item, len(data.Bids))
 	for x := range updates.Bids {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_websocket.go:365]:
```suggestion
	for x := range data.Bids {
```

File: exchanges/gateio/gateio_ws_option.go [hunk 8]
```diff
@@ -558,83 +545,68 @@ func (g *Gateio) processOptionsCandlestickPushData(data []byte) error {
...
-		snapshot := WsOptionsOrderbookSnapshot{}
-		response.Result = &snapshot
-		err := json.Unmarshal(data, &response)
+		var data WsOptionsOrderbookSnapshot
+		err := json.Unmarshal(incoming, &data)
 		if err != nil {
 			return err
 		}
-		pair, err := currency.NewPairFromString(snapshot.Contract)
+		pair, err := currency.NewPairFromString(data.Contract)
```
Comment (thrasher-) [inline exchanges/gateio/gateio_ws_option.go:565]: Optional: Could set Contract to type currency.Pair and have it Unmarshal it for you as empty pair checks happen later down the line in orderbook processing

File: exchanges/gateio/gateio_websocket.go [hunk 9]
```diff
@@ -431,41 +416,31 @@ func (g *Gateio) processOrderbookSnapshot(data []byte) error {
...
 		if err != nil {
 			return err
 		}
-		bases.Asks[x] = orderbook.Item{
-			Amount: amount,
-			Price:  price,
-		}
 	}
+	bases.Bids = make([]orderbook.Item, len(data.Bids))
 	for x := range bases.Bids {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_websocket.go:435]:
```suggestion
	for x := range data.Bids {
```

Even though all these will technically work due to `make` and ranging over it, better to be precise 😄

**Review by thrasher- — CHANGES_REQUESTED** [review]
Some of these are optional, then looks good to go!

File: exchanges/gateio/gateio_ws_futures.go (hunk 3 above)
Comment (thrasher-) [inline exchanges/gateio/gateio_ws_futures.go:530]: One guy here

File: exchanges/gateio/gateio_ws_futures.go [hunk 10]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
-		snapshot := &WsFuturesOrderbookSnapshot{}
-		response.Result = snapshot
-		err := json.Unmarshal(data, &response)
+		var data WsFuturesOrderbookSnapshot
+		err := json.Unmarshal(incoming, &data)
 		if err != nil {
 			return err
 		}
-		pair, err := currency.NewPairFromString(snapshot.Contract)
+		pair, err := currency.NewPairFromString(data.Contract)
```
Comment (thrasher-) [inline exchanges/gateio/gateio_ws_futures.go:581]: Could rid this one too

File: exchanges/gateio/gateio_ws_futures.go [hunk 11]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
 			Exchange:        g.Name,
 			Pair:            pair,
-			LastUpdated:     snapshot.TimestampInMs.Time(),
+			LastUpdated:     data.TimestampInMs.Time(),
 			VerifyOrderbook: g.CanVerifyOrderbook,
 		}
-		base.Bids = make([]orderbook.Item, len(snapshot.Bids))
-		base.Asks = make([]orderbook.Item, len(snapshot.Asks))
+		base.Asks = make([]orderbook.Item, len(data.Asks))
 		for x := range base.Asks {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_ws_futures.go:593]:
```suggestion
		for x := range data.Asks {
```

File: exchanges/gateio/gateio_ws_futures.go [hunk 12]
```diff
@@ -514,139 +511,120 @@ func (g *Gateio) processFuturesCandlesticks(data []byte, assetType asset.Item) e
...
-			base.Asks[x] = orderbook.Item{
-				Amount: snapshot.Asks[x].Size,
-				Price:  snapshot.Asks[x].Price,
-			}
+			base.Asks[x].Amount = data.Asks[x].Size
+			base.Asks[x].Price = data.Asks[x].Price
 		}
+		base.Bids = make([]orderbook.Item, len(data.Bids))
 		for x := range base.Bids {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_ws_futures.go:598]:
```suggestion
		for x := range data.Bids {
```

**Review by thrasher- — CHANGES_REQUESTED** [review]
Thanks for making those changes! After these, should be GTG 🚀

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes!

**Pull request merged**